text/2022-02-14-link-application-metrics-with-chaos-experiments.md

```diff
@@ -0,0 +1,212 @@
+
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+- How the feature is used. -->
+
+### Overview
+
+This feature is all about Chaos Dashboard, would not affect anything with other
+components and CRD.
+
+We would use Grafana as the source of application metrics, and use Grafana Panel
+as the minimum unit to display the metrics. Grafana is the widely used in
+monitoring stack, and it's a powerful design platform to make application's
+metrics looks better.
+
+And there are also other tools/platforms like Grafana, we would talk later in
+[Alternatives](#alternatives).
+
+This proposal would using Grafana Panel as the implementation, but it would keep
+the design for other platforms.
+
+The basic outcome features are:
+
+- import Grafana Panel, display it in Chaos Dashboard
+- link Grafana Panel with Chaos Experiment/Schedule/Workflow
+- enhancement ths display for metrics with chaos experiment events
+- share/reuse Grafana Panel with any user in Chaos Dashboard
+
+### Basic Concepts
+
+#### Panel
+
+Panel is the basic unit to display, **it always shown as a line chart on UI**.
+Panel could be created with outside Grafana Panel, and it should be persisted
+into the storage of Chaos Dashboard. Panel could be shared with any user, any
+chaos experiment/schedule/workflow.
+
+Panel could be linked with a chaos experiment/schedule/workflow, the relation of
+linking should be persisted.
+
+A Panel must contain some required information about its "raw panel" (the origin
+Grafana Panel), which requires:
+
+- Identify the Grafana Panel
+  - Which Grafana "backend" to use
+  - Dashboard ID of Grafana Dashboard which origin Grafana Panel belongs to
+  - Panel ID in Grafana Dashboard
+- Variables when drawing the chart
+  - Which Datasource to use
+  - Related Variables for templating "Expressions" (for querying time-series
+    data) in origin Grafana Panel
+
+Once a Panel is created, it should contains anything required to draw the chart.
+and it could be drawn on frontend immediately.
+
+> Features like `variable` and `templating expr` in Grafana is not considered in
+> short term, once the panel is created, it is "static", the charts drawn with
+> its certain parameters would not be changed (expected time range).
+
+A Panel should also contain several additional fields(for easy-to-use purposes):
+
+- Name: custom name of the panel
+- Description: description of the panel
+- Labels: panel could be labeled with many labels, it's powerful to filter the
+  panel
+
+#### Annotations
+
+Annotations are markers on line chart, it could be "instant" or "durable". It
+stands for chaos experiments, events and health check(in the future).
+
+The data of annotations would not be persisted into the storage of Chaos
+Dashboard.
+
+### Available Behaviors for Users
+
+Users could perform following behaviors:
+
+#### Use to a Grafana as "backend"
```

Comment:
Did you intend to mean `Use Grafana as "backend"`?

Reply:
I think yes. It's a typo.

Reply:
fixed!